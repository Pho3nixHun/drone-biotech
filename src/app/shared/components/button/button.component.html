@let vM = vm();

<button
    class="btn flex flex-row justify-center"
    [disabled]="disabled()"
    [ngClass]="{
        'btn-primary': !vM.secondary,
        'btn-secondary': vM.secondary,
        'btn-outline': vM.variant === 'outline',
        'btn-text': vM.variant === 'ghost',
        'btn-active': active(),
        'btn-disabled': disabled(),
    }"
>
    <ng-content select="[span, mat-icon]" />
</button>
