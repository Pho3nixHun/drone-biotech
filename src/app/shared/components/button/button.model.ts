import { WithMatIcon } from '@interfaces/mat-icon.enum';
import { WithRouterLink } from '@interfaces/with-router-link.interface';
import { WithTextNode } from '@interfaces/with-text-node.interface';

export type State = 'active' | 'disabled' | 'default';
type Variant = 'fill' | 'ghost' | 'outline';

export interface BaseButtonVM {
    variant: Variant;
    state?: State;
    secondary?: boolean;
}

interface ButtonVMWithTextNode extends BaseButtonVM, WithTextNode {
    type: 'withText';
}
interface ButtonVMWithMatIcon extends BaseButtonVM, WithMatIcon {
    type: 'withIcon';
}

export type ButtonVM = ButtonVMWithTextNode | ButtonVMWithMatIcon;

export type ButtonVMWithRouterLink = ButtonVM & WithRouterLink;
