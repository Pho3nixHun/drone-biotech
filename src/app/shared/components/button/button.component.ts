import { Component, input } from '@angular/core';
import { ButtonVM } from './button.model';
import { NgClass } from '@angular/common';

@Component({
    selector: 'app-button',
    imports: [NgClass],
    templateUrl: './button.component.html',
})
export class ButtonComponent {
    public readonly vm = input.required<ButtonVM>();
    public readonly disabled = input<boolean>();
    public readonly active = input<boolean>();
}
