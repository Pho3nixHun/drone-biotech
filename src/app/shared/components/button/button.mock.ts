import { MatIcon } from '@interfaces/mat-icon.enum';
import { ButtonVM } from './button.model';

export const enMock = {
    text: 'txt',
};

export const mockPrimaryButton: ButtonVM = {
    type: 'withText',
    textKey: enMock.text,
    variant: 'fill',
};

export const mockSecondaryButton: ButtonVM = {
    type: 'withText',
    textKey: enMock.text,
    variant: 'outline',
    secondary: true,
};

export const mockFillButton: ButtonVM = {
    type: 'withText',
    textKey: enMock.text,
    variant: 'ghost',
};
export const mockGhostButton: ButtonVM = {
    type: 'withText',
    textKey: enMock.text,
    variant: 'ghost',
};
export const mockOutlineButton: ButtonVM = {
    type: 'withText',
    textKey: enMock.text,
    variant: 'outline',
};

export const mockTextButton: ButtonVM = {
    type: 'withText',
    textKey: enMock.text,
    variant: 'outline',
};
export const mockIconButton: ButtonVM = {
    type: 'withIcon',
    icon: MatIcon.ADD,
    variant: 'outline',
};
