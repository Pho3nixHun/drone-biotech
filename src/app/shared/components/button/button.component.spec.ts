import { ComponentFixture, TestBed } from '@angular/core/testing';
import { ButtonComponent } from './button.component';
import { Component, input } from '@angular/core';
import { ButtonXVM } from './button.model';
import {
    mockFillButton,
    mockGhostButton,
    mockOutlineButton,
    mockPrimaryButton,
    mockSecondaryButton,
} from './button.mock';
import { MatIconModule } from '@angular/material/icon';

@Component({
    imports: [ButtonComponent, MatIconModule],
    template: `
        <app-button [vm]="vm()" [disabled]="disabled()" [active]="active()">
            text
            <mat-icon>search</mat-icon>
        </app-button>
    `,
})
class TestHostComponent {
    public vm = input.required<ButtonXVM>();
    public disabled = input<boolean>();
    public active = input<boolean>();
}
describe('ButtonComponent', () => {
    let fixture: ComponentFixture<TestHostComponent>;
    let compiled: HTMLElement;

    beforeEach(async () => {
        await TestBed.configureTestingModule({
            imports: [TestHostComponent],
        }).compileComponents();

        fixture = TestBed.createComponent(TestHostComponent);
        compiled = fixture.debugElement.nativeElement;
    });

    // Snapshot testing
    it('should render a primary button', () => {
        // Arrange
        fixture.componentRef.setInput('vm', mockPrimaryButton);

        // Act
        fixture.detectChanges();

        // Assert
        expect(compiled).toMatchSnapshot();
    });

    // Snapshot testing
    it('should render a secondary button', () => {
        // Arrange
        fixture.componentRef.setInput('vm', mockSecondaryButton);

        // Act
        fixture.detectChanges();

        // Assert
        expect(compiled).toMatchSnapshot();
    });

    // Snapshot testing
    it('should render a filled button', () => {
        // Arrange
        fixture.componentRef.setInput('vm', mockFillButton);

        // Act
        fixture.detectChanges();

        // Assert
        expect(compiled).toMatchSnapshot();
    });
    // Snapshot testing
    it('should render a ghost button', () => {
        // Arrange
        fixture.componentRef.setInput('vm', mockGhostButton);

        // Act
        fixture.detectChanges();

        // Assert
        expect(compiled).toMatchSnapshot();
    });

    // Snapshot testing
    it('should render an outline button', () => {
        // Arrange
        fixture.componentRef.setInput('vm', mockOutlineButton);

        // Act
        fixture.detectChanges();

        // Assert
        expect(compiled).toMatchSnapshot();
    });

    // Snapshot testing
    it('should render an active button', () => {
        // Arrange
        fixture.componentRef.setInput('vm', mockPrimaryButton);
        fixture.componentRef.setInput('active', true);

        // Act
        fixture.detectChanges();

        // Assert
        expect(compiled).toMatchSnapshot();
    });
    // Snapshot testing
    it('should render a disabled button', () => {
        // Arrange
        fixture.componentRef.setInput('vm', mockPrimaryButton);
        fixture.componentRef.setInput('disabled', true);

        // Act
        fixture.detectChanges();

        // Assert
        expect(compiled).toMatchSnapshot();
    });
});
