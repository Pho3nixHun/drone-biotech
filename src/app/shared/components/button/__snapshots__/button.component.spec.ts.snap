// Jest Snapshot v1, https://jestjs.io/docs/snapshot-testing

exports[`ButtonComponent should render a disabled button 1`] = `
<div
  id="root6"
  ng-version="20.1.7"
>
  <app-button>
    <button
      class="btn flex flex-row justify-center btn-primary btn-disabled"
      disabled=""
    >
      <span>
        text
      </span>
      <mat-icon
        aria-hidden="true"
        class="mat-icon notranslate material-icons mat-ligature-font mat-icon-no-color"
        data-mat-icon-type="font"
        role="img"
      >
        search
      </mat-icon>
    </button>
  </app-button>
</div>
`;

exports[`ButtonComponent should render a filled button 1`] = `
<div
  id="root2"
  ng-version="20.1.7"
>
  <app-button>
    <button
      class="btn flex flex-row justify-center btn-primary"
    >
      <span>
        text
      </span>
      <mat-icon
        aria-hidden="true"
        class="mat-icon notranslate material-icons mat-ligature-font mat-icon-no-color"
        data-mat-icon-type="font"
        role="img"
      >
        search
      </mat-icon>
    </button>
  </app-button>
</div>
`;

exports[`ButtonComponent should render a ghost button 1`] = `
<div
  id="root3"
  ng-version="20.1.7"
>
  <app-button>
    <button
      class="btn flex flex-row justify-center btn-primary btn-text"
    >
      <span>
        text
      </span>
      <mat-icon
        aria-hidden="true"
        class="mat-icon notranslate material-icons mat-ligature-font mat-icon-no-color"
        data-mat-icon-type="font"
        role="img"
      >
        search
      </mat-icon>
    </button>
  </app-button>
</div>
`;

exports[`ButtonComponent should render a primary button 1`] = `
<div
  id="root0"
  ng-version="20.1.7"
>
  <app-button>
    <button
      class="btn flex flex-row justify-center btn-primary"
    >
      <span>
        text
      </span>
      <mat-icon
        aria-hidden="true"
        class="mat-icon notranslate material-icons mat-ligature-font mat-icon-no-color"
        data-mat-icon-type="font"
        role="img"
      >
        search
      </mat-icon>
    </button>
  </app-button>
</div>
`;

exports[`ButtonComponent should render a secondary button 1`] = `
<div
  id="root1"
  ng-version="20.1.7"
>
  <app-button>
    <button
      class="btn flex flex-row justify-center btn-secondary btn-outline"
    >
      <span>
        text
      </span>
      <mat-icon
        aria-hidden="true"
        class="mat-icon notranslate material-icons mat-ligature-font mat-icon-no-color"
        data-mat-icon-type="font"
        role="img"
      >
        search
      </mat-icon>
    </button>
  </app-button>
</div>
`;

exports[`ButtonComponent should render an active button 1`] = `
<div
  id="root5"
  ng-version="20.1.7"
>
  <app-button>
    <button
      class="btn flex flex-row justify-center btn-primary btn-active"
    >
      <span>
        text
      </span>
      <mat-icon
        aria-hidden="true"
        class="mat-icon notranslate material-icons mat-ligature-font mat-icon-no-color"
        data-mat-icon-type="font"
        role="img"
      >
        search
      </mat-icon>
    </button>
  </app-button>
</div>
`;

exports[`ButtonComponent should render an outline button 1`] = `
<div
  id="root4"
  ng-version="20.1.7"
>
  <app-button>
    <button
      class="btn flex flex-row justify-center btn-primary btn-outline"
    >
      <span>
        text
      </span>
      <mat-icon
        aria-hidden="true"
        class="mat-icon notranslate material-icons mat-ligature-font mat-icon-no-color"
        data-mat-icon-type="font"
        role="img"
      >
        search
      </mat-icon>
    </button>
  </app-button>
</div>
`;
